### PR TITLE
[quantization] Fix `truthfulqa`

### PR DESCRIPTION
**Pull request by stamalakhov**
This PR fixes `truthfulqa` evaluation.

<details> <summary> 16-bit run of HuggingFaceTB/SmolLM2-135M-Instruct </summary>

```
Original RESULTS ARE:
|    Tasks     |Version|Filter|n-shot|  Metric   |   | Value |   |Stderr|
|--------------|------:|------|-----:|-----------|---|------:|---|-----:|
|truthfulqa_gen|      3|none  |     0|bleu_acc   |↑  | 0.2913|±  |0.0159|
|              |       |none  |     0|bleu_diff  |↑  |-5.2094|±  |0.6062|
|              |       |none  |     0|bleu_max   |↑  |21.3007|±  |0.7145|
|              |       |none  |     0|rouge1_acc |↑  | 0.2938|±  |0.0159|
|              |       |none  |     0|rouge1_diff|↑  |-7.5999|±  |0.6972|
|              |       |none  |     0|rouge1_max |↑  |46.2006|±  |0.8337|
|              |       |none  |     0|rouge2_acc |↑  | 0.2362|±  |0.0149|
|              |       |none  |     0|rouge2_diff|↑  |-8.3506|±  |0.7790|
|              |       |none  |     0|rouge2_max |↑  |29.6729|±  |0.9197|
|              |       |none  |     0|rougeL_acc |↑  | 0.2827|±  |0.0158|
|              |       |none  |     0|rougeL_diff|↑  |-7.4784|±  |0.6796|
|              |       |none  |     0|rougeL_max |↑  |43.0500|±  |0.8333|
|truthfulqa_mc1|      2|none  |     0|acc        |↑  | 0.2399|±  |0.0149|
|truthfulqa_mc2|      3|none  |     0|acc        |↑  | 0.4178|±  |0.0148|
```

```
Quantized RESULTS ARE:
|    Tasks     |Version|Filter|n-shot|  Metric   |   | Value |   |Stderr|
|--------------|------:|------|-----:|-----------|---|------:|---|-----:|
|truthfulqa_gen|      3|none  |     0|bleu_acc   |↑  | 0.3121|±  |0.0162|
|              |       |none  |     0|bleu_diff  |↑  |-3.3805|±  |0.5129|
|              |       |none  |     0|bleu_max   |↑  |16.8512|±  |0.6326|
|              |       |none  |     0|rouge1_acc |↑  | 0.3305|±  |0.0165|
|              |       |none  |     0|rouge1_diff|↑  |-5.3833|±  |0.6805|
|              |       |none  |     0|rouge1_max |↑  |40.3370|±  |0.8086|
|              |       |none  |     0|rouge2_acc |↑  | 0.2277|±  |0.0147|
|              |       |none  |     0|rouge2_diff|↑  |-5.7747|±  |0.7057|
|              |       |none  |     0|rouge2_max |↑  |23.0924|±  |0.8486|
|              |       |none  |     0|rougeL_acc |↑  | 0.3121|±  |0.0162|
|              |       |none  |     0|rougeL_diff|↑  |-5.3809|±  |0.6535|
|              |       |none  |     0|rougeL_max |↑  |37.0933|±  |0.7974|
|truthfulqa_mc1|      2|none  |     0|acc        |↑  | 0.2485|±  |0.0151|
|truthfulqa_mc2|      3|none  |     0|acc        |↑  | 0.4126|±  |0.0148|
```
which are pretty similar

log: 
[log_SMOLLM_eval_9.txt](https://github.com/user-attachments/files/26704282/log_SMOLLM_eval_9.txt)

</details>

<details> <summary>Run on 4-bit `smse` on unsloth/Llama-3.2-3B-Instruct </summary>

```
Original RESULTS ARE:
|    Tasks     |Version|Filter|n-shot|  Metric   |   | Value |   |Stderr|
|--------------|------:|------|-----:|-----------|---|------:|---|-----:|
|truthfulqa_gen|      3|none  |     0|bleu_acc   |↑  | 0.6193|±  |0.0170|
|              |       |none  |     0|bleu_diff  |↑  |16.4238|±  |1.1327|
|              |       |none  |     0|bleu_max   |↑  |35.2506|±  |0.8846|
|              |       |none  |     0|rouge1_acc |↑  | 0.5998|±  |0.0172|
|              |       |none  |     0|rouge1_diff|↑  |23.1605|±  |1.6231|
|              |       |none  |     0|rouge1_max |↑  |59.2485|±  |1.0459|
|              |       |none  |     0|rouge2_acc |↑  | 0.5349|±  |0.0175|
|              |       |none  |     0|rouge2_diff|↑  |24.4049|±  |1.6812|
|              |       |none  |     0|rouge2_max |↑  |47.2195|±  |1.2588|
|              |       |none  |     0|rougeL_acc |↑  | 0.5887|±  |0.0172|
|              |       |none  |     0|rougeL_diff|↑  |22.9323|±  |1.6371|
|              |       |none  |     0|rougeL_max |↑  |57.5084|±  |1.0770|
|truthfulqa_mc1|      2|none  |     0|acc        |↑  | 0.3366|±  |0.0165|
|truthfulqa_mc2|      3|none  |     0|acc        |↑  | 0.5147|±  |0.0150|

Quantized RESULTS ARE:
|    Tasks     |Version|Filter|n-shot|  Metric   |   | Value |   |Stderr|
|--------------|------:|------|-----:|-----------|---|------:|---|-----:|
|truthfulqa_gen|      3|none  |     0|bleu_acc   |↑  | 0.5508|±  |0.0174|
|              |       |none  |     0|bleu_diff  |↑  |12.0819|±  |1.0953|
|              |       |none  |     0|bleu_max   |↑  |33.8811|±  |0.8637|
|              |       |none  |     0|rouge1_acc |↑  | 0.5177|±  |0.0175|
|              |       |none  |     0|rouge1_diff|↑  |17.0869|±  |1.5387|
|              |       |none  |     0|rouge1_max |↑  |58.7352|±  |0.9873|
|              |       |none  |     0|rouge2_acc |↑  | 0.4835|±  |0.0175|
|              |       |none  |     0|rouge2_diff|↑  |17.0286|±  |1.6394|
|              |       |none  |     0|rouge2_max |↑  |45.4835|±  |1.1998|
|              |       |none  |     0|rougeL_acc |↑  | 0.5080|±  |0.0175|
|              |       |none  |     0|rougeL_diff|↑  |16.9061|±  |1.5483|
|              |       |none  |     0|rougeL_max |↑  |56.7794|±  |1.0144|
|truthfulqa_mc1|      2|none  |     0|acc        |↑  | 0.3097|±  |0.0162|
|truthfulqa_mc2|      3|none  |     0|acc        |↑  | 0.4869|±  |0.0148|
```

</details>

<details> <summary> Run 4-bit mse on unsloth/Llama-3.2-3B-Instruct </summary>

```
Original RESULTS ARE:
|    Tasks     |Version|Filter|n-shot|  Metric   |   | Value |   |Stderr|
|--------------|------:|------|-----:|-----------|---|------:|---|-----:|
|truthfulqa_gen|      3|none  |     0|bleu_acc   |↑  | 0.6193|±  |0.0170|
|              |       |none  |     0|bleu_diff  |↑  |16.4238|±  |1.1327|
|              |       |none  |     0|bleu_max   |↑  |35.2506|±  |0.8846|
|              |       |none  |     0|rouge1_acc |↑  | 0.5998|±  |0.0172|
|              |       |none  |     0|rouge1_diff|↑  |23.1605|±  |1.6231|
|              |       |none  |     0|rouge1_max |↑  |59.2485|±  |1.0459|
|              |       |none  |     0|rouge2_acc |↑  | 0.5349|±  |0.0175|
|              |       |none  |     0|rouge2_diff|↑  |24.4049|±  |1.6812|
|              |       |none  |     0|rouge2_max |↑  |47.2195|±  |1.2588|
|              |       |none  |     0|rougeL_acc |↑  | 0.5887|±  |0.0172|
|              |       |none  |     0|rougeL_diff|↑  |22.9323|±  |1.6371|
|              |       |none  |     0|rougeL_max |↑  |57.5084|±  |1.0770|
|truthfulqa_mc1|      2|none  |     0|acc        |↑  | 0.3366|±  |0.0165|
|truthfulqa_mc2|      3|none  |     0|acc        |↑  | 0.5147|±  |0.0150|

Quantized RESULTS ARE:
|    Tasks     |Version|Filter|n-shot|  Metric   |   | Value |   |Stderr|
|--------------|------:|------|-----:|-----------|---|------:|---|-----:|
|truthfulqa_gen|      3|none  |     0|bleu_acc   |↑  | 0.4835|±  |0.0175|
|              |       |none  |     0|bleu_diff  |↑  | 7.0149|±  |1.0841|
|              |       |none  |     0|bleu_max   |↑  |30.4182|±  |0.8508|
|              |       |none  |     0|rouge1_acc |↑  | 0.4749|±  |0.0175|
|              |       |none  |     0|rouge1_diff|↑  |10.5110|±  |1.4802|
|              |       |none  |     0|rouge1_max |↑  |55.5551|±  |0.9786|
|              |       |none  |     0|rouge2_acc |↑  | 0.4113|±  |0.0172|
|              |       |none  |     0|rouge2_diff|↑  | 9.6823|±  |1.5941|
|              |       |none  |     0|rouge2_max |↑  |40.8478|±  |1.1862|
|              |       |none  |     0|rougeL_acc |↑  | 0.4614|±  |0.0175|
|              |       |none  |     0|rougeL_diff|↑  |10.1191|±  |1.4909|
|              |       |none  |     0|rougeL_max |↑  |53.2202|±  |1.0071|
|truthfulqa_mc1|      2|none  |     0|acc        |↑  | 0.2815|±  |0.0157|
|truthfulqa_mc2|      3|none  |     0|acc        |↑  | 0.4652|±  |0.0148|

```

</details>

<details> <summary>./ccex test --include-internal</summary>

```
...
----------------------------------------------------------------------
Ran 1099 tests in 532.613s

OK (skipped=20)
```
[test_log.txt](https://github.com/user-attachments/files/26667845/test_log.txt)

</details>

Related: #617 
TICO-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 1]
```diff
@@ -187,7 +186,16 @@ def quantize_using_PTQ(q_m, calib_inputs, args):
...
         for inp in tqdm.tqdm(calib_inputs):
-            q_m(inp.to(device))
+            if args.calibrate_use_cache:
+                outputs = q_m(inp[..., :-1].to(device), use_cache=True)
+                # TODO add padding?
+                q_m(
+                    inp[..., -1:].to(device),
+                    past_key_values=outputs.past_key_values,
+                    use_cache=True,
+                )
```
Comment (stamalakhov) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py:229]: We need to turn on `use_cache` option to calibrate kv-cache observers (both input and output), that is why prefill-decode model is used. First run produces kv-cache, second run uses it on a single token inference.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 2]
```diff
@@ -341,6 +349,12 @@ def main():
         type=str,
         default=None,
     )
+    parser.add_argument(
+        "--calibrate_use_cache",
```
Comment (stamalakhov) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py:426]: not sure about the option name, bit it should be used only in PTQ quantizer calibration. 
Any other suggestions are welcome.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 3]
```diff
@@ -383,7 +397,7 @@ def main():
     else:
         print("Skipping SpinQuant preprocessing …")
 
-    model.config.use_cache = False  # TODO use args for it
+    model.config.use_cache = False
```
Comment (stamalakhov) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py:473]: it occured that GPTQ is not very kv-cache friendly, as other possible techniques. Right now it's needed only in PTQ.

File: tico/quantization/wrapq/wrappers/llama/quant_attention.py [hunk 4]
```diff
@@ -205,10 +205,16 @@ def _concat_kv(
         if past is None:
             return k_new, v_new
         past_k, past_v = past
-        past_k = self._fq(past_k, self.obs_past_key)
-        past_v = self._fq(past_v, self.obs_past_value)
-        k = torch.cat([past_k[:, kv_idx, :, :], k_new], dim=1)
-        v = torch.cat([past_v[:, kv_idx, :, :], v_new], dim=1)
+        if past_k is not None:
```
Comment (stamalakhov) [inline tico/quantization/wrapq/wrappers/llama/quant_attention.py:208]: `DynamicCache` tuple can be None, but the values of the tuple can be None.

File: tico/quantization/wrapq/wrappers/llama/quant_model.py [hunk 5]
```diff
@@ -143,25 +143,83 @@ def __init__(
...
+            if attention_mask.dtype == torch.int64:
+                attention_mask = attention_mask == 1  # convert to bool
+            mask = self.causal_mask_template[..., :q_len, :k_len].to(device)
+            # only padding which is assumed to change causal_mask
+            additive = torch.zeros_like(attention_mask, dtype=torch.float32)
+            additive = additive.masked_fill(~attention_mask, float("-120"))
+            mask = torch.max(
+                torch.tensor(float("-120")).to(device), additive + mask
+            )  # so -120-120->-120, -120+0->-120, 0-120->-120, 0+0->0
+            return mask
```
Comment (stamalakhov) [inline tico/quantization/wrapq/wrappers/llama/quant_model.py]: `lm_eval ` uses `int64` attention_mask with possible padding, so we need to AND usual casual_mask and external attention_mask with possible padding.

File: tico/quantization/wrapq/wrappers/llama/quant_model.py [hunk 6]
```diff
@@ -202,52 +260,68 @@ def forward(
...
-            )
-            cache_position = torch.arange(
-                past_seen_tokens,
-                past_seen_tokens + inputs_embeds.shape[1],
-                device=inputs_embeds.device,
-            )
-
-        if position_ids is None:
-            position_ids = cache_position.unsqueeze(0)
+            past_key_values = []
```
Comment (stamalakhov) [inline tico/quantization/wrapq/wrappers/llama/quant_model.py:268]: `DynamicCache` seems to be non-exportable to circle, so let's just use list instead.

File: tico/quantization/wrapq/wrappers/llama/quant_model.py [hunk 7]
```diff
@@ -271,7 +355,11 @@ def forward(
 
         output = BaseModelOutputWithPast(
             last_hidden_state=hidden_states,
-            past_key_values=past_key_values if use_cache else None,
+            past_key_values=(
+                past_key_values
+                if use_cache and isinstance(past_key_values, DynamicCache)
+                else present_key_values
+            ),
```
Comment (stamalakhov) [inline tico/quantization/wrapq/wrappers/llama/quant_model.py:367]: in case DynamicCache is used it was sent externally for non-export purposes, so let's return updated kv-cache.
otherwise return new values to be managed by runtime.

File: tico/quantization/wrapq/wrappers/llama/quant_model_for_causal_lm.py [hunk 8]
```diff
@@ -99,6 +107,9 @@ def forward(
         output_attentions = self.config.output_attentions
         output_hidden_states = self.config.output_hidden_states
         return_dict = self.config.use_return_dict
+        if "return_dict" in kwargs:
+            # lm_eval set return_dict explicitely in kwargs
+            return_dict = kwargs.pop("return_dict")
```
Comment (stamalakhov) [inline tico/quantization/wrapq/wrappers/llama/quant_model_for_causal_lm.py]: it occures that `lm_eval` uses `return_dict` set explicitely in kwargs, so we need to erase it from kwargs to avoid duplication.

**Comment by stamalakhov** [comment]
@mhs4670go 
Please take another look!

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 9]
```diff
@@ -220,7 +219,16 @@ def quantize_using_PTQ(q_m, calib_inputs, args):
     device = torch.device(args.device)
     with torch.no_grad():
         for inp in tqdm.tqdm(calib_inputs):
-            q_m(inp.to(device))
+            if args.calibrate_use_cache:
+                outputs = q_m(inp[..., :-1].to(device), use_cache=True)
+                # TODO add padding?
+                q_m(
```
Comment (stamalakhov) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py:225]: @mhs4670go 
Or should it be original float model to calibrate for input kv_cache?

File: tico/quantization/wrapq/wrappers/llama/quant_model_for_causal_lm.py (hunk 8 above)
Comment (mhs4670go) [reply tico/quantization/wrapq/wrappers/llama/quant_model_for_causal_lm.py]: @stamalakhov First of all, the wrapper currently ignores the runtime `return_dict` argument and always falls back to `config.use_return_dict`, while still forwarding return_dict through `**kwargs`, which leads to duplicate argument errors.

I think the right fix is to make `QuantLlamaForCausalLM.forward` HF-compatible by explicitly accepting `return_dict`, and only falling back to `self.config` when they are None, rather than just poping out the value.

How about this one?

```python
def forward(..., return_dict=None, **kwargs):
    return_dict = return_dict if return_dict is not None else self.config.use_return_dict

    outputs = self.model(..., return_dict=return_dict, **kwargs)
```

Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_model_for_causal_lm.py]: @mhs4670go 
Thank you! Fixed.

**Comment by mhs4670go** [comment]
> Please take another look!

Sorry for the delayed review. I’ve been tied up with internal training at work, and the changes in this PR seem to require more careful consideration than expected, which caused the delay. I’ll complete the review within this week—thanks for your understanding.
